daceml/onnx/op_implementations/pure_implementations.py

```diff
@@ -0,0 +1,617 @@
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+
+        def prog(A, B, C):
+            C[:] = A - B
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="Mul", name="pure")
+class PureMul(ONNXForward):
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+
+        def prog(A, B, C):
+            C[:] = A * B
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="Div", name="pure")
+class PureDiv(ONNXForward):
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+
+        def prog(A, B, C):
+            C[:] = A / B
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="ReduceMean", name="pure")
+class PureReduceMean(ONNXForward):
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+
+        axes = node.axes
+
+        # when keepdims is true, this works but there is a useless copy. We just leave this for now; this can be fixed
+        # with a reshape node when those exist.
+        def prog(data, reduced):
+            reduced[:] = np.mean(data, axis=axes)
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="Erf", name="pure")
+class PureErf(ONNXForward):
+    @staticmethod
+    def forward_can_be_applied(node: ONNXOp, state: SDFGState,
+                               sdfg: SDFG) -> bool:
+        return in_desc_with_name(node, state, sdfg, 'input').dtype in [
+            dace.float16, dace.float32, dace.float64
+        ]
+
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        node.validate(sdfg, state)
+
+        def prog(input, output):
+            output[:] = dace.elementwise(lambda x: erf(x), input)
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="MatMul", name="pure")
+class PureMatMul(ONNXForward):
+    @staticmethod
+    def forward_can_be_applied(node: ONNXOp, state: SDFGState,
+                               sdfg: SDFG) -> bool:
+        in_edges = state.in_edges(node)
+        input0_dim = len(in_desc_with_name(node, state, sdfg, "A").shape)
+        input1_dim = len(in_desc_with_name(node, state, sdfg, "B").shape)
+
+        if input0_dim == 4 and input1_dim == 4:
+            return True
+
+        if input0_dim == 3 and input1_dim == 2:
+            return True
+
+        if input0_dim == 2 and input1_dim == 2:
+            return True
+
+        return False
+
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
```

Comment:
is there a way to make this nicer? i.e., not reuse existing code or expand to our existing library nodes?

Reply:
we now have interesting implementations of those nodes for different platforms and should use them.

Reply:
There are likely better ways, but if there's no direct way (I'm not really aware of a general equivalent to `np.matmul` in dace) then I would prefer to postpone this to a new PR